content/blog/cpu-profiling/index.md

```diff
@@ -0,0 +1,86 @@
+
+More recently, sampling-based profilers have become widely used due to their low overhead. The basic idea behind sampling-based profilers is to periodically interrupt the program and record the current stack trace. The stack trace is recovered by looking at the instruction pointer of the application on the CPU, and then inspecting the stack to find the instruction pointers of all the parent functions (frames) as well. Walking the stack to reconstruct a stack trace has some complexities, but the basic case is shown below. One starts at the leaf frame, and uses frame pointers to successively find the next parent frame. Each stack frame contains a return address instruction pointer which is recorded to construct the entire stack trace.
+
+::: div image-m
+<svg title="A program’s call stack. Frame pointers can be used to walk the stack and record the return addresses to generate a stack trace." src='callstack.png' />
+:::
+
+By sampling stack traces many thousands of times, one gets a good idea of where the code spends its time. This is fundamentally a statistical approach, so the more samples are collected, the more confidence you’ll have that you’ve found a real hot-spot. You also have to ensure there’s no correlation between your sampling methodology and the target application, so you can trust the results, but a simple timer-based approach typically works well.
+
+Sampling profilers can be very efficient, to the point that there is negligible overhead. For example, if one samples a stack trace every 10 ms, and we assume (1) the sampling process requires 50,000 instructions, and (2) that the CPU processes 5 billion instructions per second, then a rough calculation (10000*100/5 B) shows a theoretical performance overhead of only 0.1%.
+
+A third approach to profiling is simulation, as used by Valgrind/Callgrind. Valgrind uses no instrumentation, but runs your program through a virtual machine which profiles the execution. This approach provides a lot of information, but is also high in overheads.
+
+The table below summarizes properties of a number of popular performance profilers.
+
+| Profiler | Methodology | Deployment | Traces Libraries/System Calls? | Performance Overhead
+| --- | --- | --- | --- | --- |
+| [gprof](https://sourceware.org/binutils/docs/gprof/) | Instrumentation + Sampling | Recompile & Rerun | No | High (up to [260%](https://www.researchgate.net/publication/221235356_Low-overhead_call_path_profiling_of_unmodified_optimized_code)) |
+| [Callgrind](https://valgrind.org/docs/manual/cl-manual.html) | Simulation | Rerun | Yes? | Very High ([>400%](https://www.cs.cmu.edu/afs/cs.cmu.edu/project/cmt-40/Nice/RuleRefinement/bin/valgrind-3.2.0/docs/html/cl-manual.html)) |
```

Comment:
Remove the question mark on Yes?

Reply:
I removed the question mark for both Callgrind and gperftools.